netket/models/rnn.py

```diff
@@ -0,0 +1,276 @@
+    When the actual number of previous neighbors of a site is less than `max_prev_neighbors`,
+    use -1 to denote zero paddings instead of memory from a neighbor."""
+    graph: Optional[AbstractGraph] = None
+    """graph of the physical system."""
+    param_dtype: DType = jnp.float64
+    """the dtype of the computation (default: float64)."""
+    kernel_init: NNInitFunc = default_kernel_init
+    """initializer for the weights."""
+    bias_init: NNInitFunc = zeros
+    """initializer for the biases."""
+    machine_pow: int = 2
+    """exponent to normalize the outputs of `__call__`."""
+
+    def reorder(self, inputs: Array, axis: int = 0) -> Array:
+        if self.reorder_idx is None:
+            return inputs
+        else:
+            idx = jnp.asarray(self.reorder_idx)
+            return inputs.take(idx, axis)
+
+    def inverse_reorder(self, inputs: Array, axis: int = 0) -> Array:
+        if self.inv_reorder_idx is None:
+            return inputs
+        else:
+            idx = jnp.asarray(self.inv_reorder_idx)
+            return inputs.take(idx, axis)
+
+
+class _LSTMNet(RNN):
+    def setup(self):
+        features = _get_feature_list(self)
+        self._layers = [
+            LSTMLayer(
+                features=features[i],
+                exclusive=(i == 0),
+                reorder_idx=self.reorder_idx,
+                inv_reorder_idx=self.inv_reorder_idx,
+                prev_neighbors=self.prev_neighbors,
+                param_dtype=self.param_dtype,
+                kernel_init=self.kernel_init,
+                bias_init=self.bias_init,
+            )
+            for i in range(self.layers)
+        ]
+
+
+class _GRUNet1D(RNN):
+    def setup(self):
+        features = _get_feature_list(self)
+        self._layers = [
+            GRULayer1D(
+                features=features[i],
+                exclusive=(i == 0),
+                reorder_idx=self.reorder_idx,
+                inv_reorder_idx=self.inv_reorder_idx,
+                prev_neighbors=self.prev_neighbors,
+                param_dtype=self.param_dtype,
+                kernel_init=self.kernel_init,
+                bias_init=self.bias_init,
+            )
+            for i in range(self.layers)
+        ]
+
+
+def _get_inv_idx(idx):
+    idx = np.asarray(idx)
+    inv = np.empty_like(idx)
+    for i, k in enumerate(idx):
+        inv[k] = i
+    inv = HashableArray(inv)
+    return inv
+
+
+def _get_extent(graph):
+    V = graph.n_nodes
+
+    if isinstance(graph, Lattice):
+        assert len(graph.extent) == 2
+        L, M = graph.extent
+    else:
+        L = M = int(sqrt(V))
+    assert L * M == V
+
+    return V, L, M
+
+
+def _get_snake_inv_reorder_idx(graph):
+    V, L, M = _get_extent(graph)
+    idx = np.arange(V, dtype=np.intp).reshape((L, M))
+    idx[1::2, :] = idx[1::2, ::-1]
+    idx = idx.flatten()
+    idx = HashableArray(idx)
+    return idx
+
+
+def _get_inv_reorder_idx(graph):
```

Comment:
```suggestion
def _get_inv_reorder_idx(graph: AbstractGraph) -> HashableArray[int]:
```

Reply:
There may be an issue in future whether `HashableArray[int]` is compatible with `HashableArray[np.int32]`, `HashableArray[np.uint32]` and more, but for now let's just write `HashableArray[int]`

Reply:
Ah we haven't implemented `HashableArray[...]` (the error message is `TypeError: 'type' object is not subscriptable`), so let's just write `HashableArray`